service/controller/resource/ipam/azure_machinepool_subnet_collector.go

```diff
@@ -0,0 +1,190 @@
+package ipam
+
+import (
+	"context"
+	"net"
+	"sync"
+
+	"github.com/giantswarm/ipam"
+	"github.com/giantswarm/microerror"
+	"golang.org/x/sync/errgroup"
+	capzV1alpha3 "sigs.k8s.io/cluster-api-provider-azure/api/v1alpha3"
+	ctrl "sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/giantswarm/azure-operator/v4/client"
+	"github.com/giantswarm/azure-operator/v4/pkg/helpers"
+	"github.com/giantswarm/azure-operator/v4/service/controller/key"
+)
+
+type AzureMachinePoolSubnetCollectorConfig struct {
+	AzureClientFactory *client.Factory
+	Client             ctrl.Client
+}
+
+// AzureMachinePoolSubnetCollector is a Collector implementation that collects all subnets that are
+// already allocated in tenant cluster virtual network. See Collect function implementation and
+// docs for more details.
+type AzureMachinePoolSubnetCollector struct {
+	azureClientFactory *client.Factory
+	client             ctrl.Client
+}
+
+func NewAzureMachineSubnetCollector(config AzureMachinePoolSubnetCollectorConfig) (*AzureMachinePoolSubnetCollector, error) {
+	if config.AzureClientFactory == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.AzureClientFactory must not be empty", config)
+	}
+	if config.Client == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Client must not be empty", config)
+	}
+
+	c := &AzureMachinePoolSubnetCollector{
+		azureClientFactory: config.AzureClientFactory,
+		client:             config.Client,
+	}
+
+	return c, nil
+}
+
+// Collect function returns all subnets that are already allocated in tenant cluster virtual
+// network. These include subnets set in AzureCluster CR and all subnets that are created in tenant
+// cluster's Azure virtual network.
+//
+// Why getting both of these?
+//   - Subnets from AzureCluster CR: these are desired subnets for the tenant cluster, they might
+//     be already deployed in Azure or not.
+//   - Subnets in Azure virtual network: In addition to subnets from AzureCluster CR that should be
+//     eventually deployed here, there might be some other subnets that are created outside of
+//     tenant cluster. For existing pre-node-pool clusters, legacy subnets, if they still exist,
+//     will be collected here.
+func (c *AzureMachinePoolSubnetCollector) Collect(ctx context.Context, obj interface{}) ([]net.IPNet, error) {
+	var err error
+	var mutex sync.Mutex
+	var reservedSubnets []net.IPNet
+
+	azureMachinePool, err := key.ToAzureMachinePool(obj)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	// Get AzureCluster CR where the subnets are stored.
+	azureCluster, err := helpers.GetAzureClusterFromMetadata(ctx, c.client, azureMachinePool.ObjectMeta)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
```

Comment:
Bail out if `azureCluster.Spec.NetworkSpec.Vnet.CidrBlock` is still empty.

Reply:
Done.